### ISSUE TITLE
User-provided materializer will be used to create system wide objects

### ISSUE THREAD

**Issue by samueltardieu** [akka/akka#20032]
`Http().singleRequest()` creates or reuses a cached connection stored in a ActorSystem-wide pool. However, this connection is created using the user-provided materializer. If this materializer is tied to an actor through its `context`, other actors will no longer be able to contact the same host/port after the initial materializer has been shutdown.

This is likely to happen if the user has a `implicit val materializer = ActorMaterializer()(context)` in their actor.

ActorSystem-wide cached objects should be created using a materializer with an ActorSystem scope instead of the materializer provided by the caller.

The following example illustrates the problem:

``` scala
import akka.actor.ActorSystem
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.Flow
import org.specs2.mutable._

import scala.concurrent.Await
import scala.concurrent.duration._

class MaterializerSpec extends Specification {

  implicit val system = ActorSystem("test-system")
  implicit val dispatcher = system.dispatcher
  val systemMaterializer = ActorMaterializer()(system)

  val handler = Flow[HttpRequest].map(_ => new HttpResponse())
  val bindingFuture = Http().bindAndHandle(handler, "localhost", 0)(systemMaterializer)
  val port = Await.result(bindingFuture.map(_.localAddress.getPort), 1.second)
  val uri = s"http://localhost:$port"

  "non-global materializer should not be globally stored" in {

    {
      implicit val localMaterializer = ActorMaterializer()(system)
      // The local materializer will wrongly be used to create an ActorSystem-wide connection
      // in the cached connection pool.
      val response = Await.result(Http().singleRequest(new HttpRequest(uri = uri)), 1.second)
      localMaterializer.shutdown()
      response.status.intValue() must be equalTo 200
    }
    {
      implicit val localMaterializer = ActorMaterializer()(system)
      // The following request will fail with a timeout because the local materializer of the
      // first request has been used to create the cached connection in the connection pool.
      //
      // Commenting the local materializer shutdown above will make it work as the previous
      // materializer would still be alive.

      // Using a request to another host/port will also work as a new connection will be
      // created in the connection pool with this local materializer.
      val response = Await.result(Http().singleRequest(new HttpRequest(uri = uri)), 1.second)
      localMaterializer.shutdown()
      response.status.intValue() must be equalTo 200
    }
  }
}
```

**Comment by samueltardieu** [comment]
Here is another example, based on actors, which illustrates that the user might not think they do anything wrong. The `RequestActor` below waits for "go", fetches an URL and returns the status code then terminates. As a side effect, when the first actor terminates, its materializer that has been used to create the cached connection has been shutdown.

``` scala
import akka.pattern.pipe
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.Flow
import akka.util.Timeout
import org.specs2.mutable._

import scala.concurrent.Await
import scala.concurrent.duration._

class MaterializerSpec extends Specification {

  implicit val system = ActorSystem("test-system")
  implicit val dispatcher = system.dispatcher
  val systemMaterializer = ActorMaterializer()(system)

  val handler = Flow[HttpRequest].map(_ => new HttpResponse())
  val bindingFuture = Http().bindAndHandle(handler, "localhost", 0)(systemMaterializer)
  val port = Await.result(bindingFuture.map(_.localAddress.getPort), 1.second)
  val uri = s"http://localhost:$port"

  class RequestActor extends Actor {

    implicit val materializer = ActorMaterializer()(context)

    def receive = {
      case "go" =>
        Http().singleRequest(new HttpRequest(uri = uri)).map(_.status.intValue()).pipeTo(sender).andThen {
          case _ => context.stop(self)
        }
    }
  }

  "non-global materializer should not be globally stored" in {
    implicit val timeout: Timeout = 1.second
    Await.result((system.actorOf(Props(new RequestActor)) ? "go").mapTo[Int], 1.second) should be equalTo 200
    Await.result((system.actorOf(Props(new RequestActor)) ? "go").mapTo[Int], 1.second) should be equalTo 200
    Await.result((system.actorOf(Props(new RequestActor)) ? "go").mapTo[Int], 1.second) should be equalTo 200
    Await.result((system.actorOf(Props(new RequestActor)) ? "go").mapTo[Int], 1.second) should be equalTo 200
  }
}
```

**Comment by ktoso** [comment]
The core of this is fixed via #20082 however we should look into possibly allowing configuring of the "systemMaterializer"
